http-api/src/v1/projects.rs

```diff
@@ -149,35 +149,93 @@ async fn commit_handler(
     Extension(ctx): Extension<Context>,
     Path((project, sha)): Path<(Urn, One)>,
 ) -> impl IntoResponse {
-    let reference = Reference::head(Namespace::from(project.clone()), None, sha.to_owned());
-    let commit = browse(reference, ctx.paths.clone(), |browser| {
-        let oid = browser.oid(&sha)?;
-        radicle_source::commit(browser, oid)
-    })
-    .await?;
-
     let projects = ctx.projects.read().await;
+
+    let sha = git::Oid::from_str(&sha)?;
+    let repo = git::Repository::new(ctx.paths.git_dir()).map_err(Error::from)?;
+    let mut browser = git::Browser::new_with_namespace(
+        &repo,
+        &git::Namespace::try_from(project.encode_id().as_str())
+            .map_err(|_| Error::MissingNamespace)?,
+        sha,
+    )
+    .map_err(Error::from)?;
+    browser.commit(sha)?;
+
+    let history = browser.get();
+    let commit = history.first();
+
+    let diff = if let Some(parent) = commit.parents.first() {
+        browser.diff(*parent, sha)?
+    } else {
+        browser.initial_diff(sha)?
+    };
+
+    let mut deletions = 0;
+    let mut additions = 0;
+
+    for file in &diff.modified {
```

Comment:
Something you could do to make this function a little bit terser is to write an impl:
```rust
impl From<&Diff> for CommitStats {
  fn from(diff: &Diff) -> CommitStats {
    let mut deletions = 0;
    let mut additions = 0;
    /* these for loops in here */
    Self { additions, deletions }
  }
}
```

Then here, you would replace alllll this code with:
```rust
let stats = CommitStats::from(&diff);
```

Reply:
Interesting that you say that, I think we can do even better.
We could add all this to `radicle_surf::diff::Diff` and then we only would need `Diff::stats()` for [`Changeset`](https://github.com/radicle-dev/radicle-client-services/blob/c3212beb1ce198ed31d03d7bd4b09c153900e118/http-api/src/v1/projects.rs#L685)  and for this handler. wdyt?

Reply:
Ya, that works! :) What is a `Changeset`? I'm curious as to why it has a `Diff` and a `Vec<Commit>`.

Reply:
It is a wrapper around `Diff` `Stats` and `Vec<Commit>` have everything patch related in one object.
I'd have to think about if we can maybe, by extracting the `stats` function, into its elements.

Reply:
> It is a wrapper around `Diff` `Stats` and `Vec<Commit>` to have everything patch related in one struct.

Ah, I see :)

Also, I just came across this too https://docs.rs/git2/0.15.0/git2/struct.Diff.html#method.stats

Reply:
Nice! Would it be possible, to remove our `diff::Diff` implementation and use the `git2::Diff` implementation?
Or maybe add some `TryFrom` trait between the two types?

Reply:
Ok just refactored the `stats` retrieval out [7b2c38a](https://github.com/radicle-dev/radicle-client-services/pull/196/commits/7b2c38a40245b6c121068547b1d6a1b22b84ee1c)
and moved it to `radicle_surf` https://github.com/radicle-dev/radicle-git/pull/11

Reply:
> Nice! Would it be possible, to remove our `diff::Diff` implementation and use the `git2::Diff` implementation?
> Or maybe add some `TryFrom` trait between the two types?

There is a `TryFrom` :sweat_smile: https://github.com/radicle-dev/radicle-git/blob/main/radicle-surf/src/diff/git.rs#L86

Reply:
Nice, but was more thinking along the lines of directly using `git2::Diff` but it seems like to have our own `diff::Diff` allows us some more space for implementation.

Reply:
Ya, this is true :) We'll have to see what happens when we look at the `radicle-surf` re-design.